Add Functionality Allowing Super Admins to Confirm/Unconfirm User Emails

aaronskiba opened the pull request:
Fixes #1006

Changes proposed in this PR:
- Include email confirmation status as an editable attribute on the `/super_admin/users/:id` path.
  ![Screenshot from 2025-02-10 12-19-02](https://github.com/user-attachments/assets/d58ff379-9f9c-4263-829f-1d32f23f1e76)
  - The added logic includes special handling for superadmins. Specifically, it prevents the un-confirming of any already confirmed super admin email.
- This added feature helps address an encountered issue (https://github.com/portagenetwork/roadmap/issues/1005) where some institutions were blocking emails from `@portagenetwork.ca`, and as a result, preventing users from receiving email confirmation instructions.

aaronskiba committed:
Add logic allowing superusers to confirm emails

This change addresses an issue where some users reported not receiving the confirmation instructions email. Rather than having to manually update the users email confirmation status within the Rails console, this change allows users to confirm/unconfirm user emails within the app.

aaronskiba committed:
Add users_controller_spec + confirmation tests

aaronskiba committed:
Fix email confirmation checkbox's alignment

aaronskiba committed:
Prevent unconfirming of super admin emails

aaronskiba committed:
Prevent superadmin un-confirmation if UI bypassed

`spec/controllers/super_admin/users_controller_spec.rb`:
- Added test to prevent un-confirming of any super admin email within app
`app/controllers/super_admin/users_controller.rb`:
- Added !@user.can_super_admin? check to enable the aforementioned test to pass
- Also, updated a `patch` statement to `put`

aaronskiba committed:
Update CHANGELOG.md

momo3404 commented on app/controllers/super_admin/users_controller.rb:
```diff
@@ -135,5 +137,20 @@ def merge_accounts
+
+    def handle_confirmed_at_param(attrs)
+      # if an unconfirmed email is now being confirmed
+      if !@user.confirmed? && attrs[:confirmed_at] == '1'
```
It could be useful to leave a comment stating that 1/0 refer to the form

momo3404 commented on app/controllers/super_admin/users_controller.rb:
```diff
@@ -135,5 +137,20 @@ def merge_accounts
+      else
+        # else delete the param
+        # (keeps value nil for unconfirmed user and maintains previous Time value for confirmed user)
+        attrs.delete(:confirmed_at)
```
Just curious, how would deleting the 'confirmed_at' parameter keep the value nil or maintain the previous time value?

aaronskiba replied:
Good question. Let's look at lines 41-43 of this file:

```ruby
attrs = handle_confirmed_at_param(attrs)

if @user.update(attrs)
```

If `attrs.delete(:confirmed_at)` is executed within `def handle_confirmed_at_param(attrs)`, then `@user.confirmed_at` shouldn't be affected by `@user.update(attrs)`.

I tried to write `def handle_confirmed_at_param` so that `@user.confirmed_at` is only updated when a confirmed_status changes from unconfirmed -> confirmed or vice versa. So, for example, if a user is already confirmed, instead of updating the @user.confirmed_at value to `Time.current`, clicking save while the box is checked will have no effect.

momo3404 replied:
Makes sense, thanks for the clarification!

momo3404 reviewed (dismissed):
Test file ran successfully and confirmed_at parameter seems to work properly

aaronskiba committed:
Add comment to `def handle_confirmed_at_param`

aaronskiba committed:
Clarify email confirmation checkbox instructions

Improves UX and reduces confusion when updating a user's
email confirmation status.

- Moved confirmation status inline with the "Email status:" heading
- Updated checkbox helper text to reflect current @user.confirmed? state:
  - Shows "(Check to confirm email)" if unconfirmed
  - Shows "(Uncheck to unconfirm email)" if confirmed

momo3404 commented on app/views/super_admin/users/_email_confirmation_status.html.erb:
```diff
@@ -0,0 +1,12 @@
+                                     style: 'vertical-align: middle;
+                                             margin-top: -2px;' }) %>
+    <% checkbox_helper_text = @user.confirmed? ? _('(Uncheck to unconfirm email)') : _('(Check to confirm email)') %>
+    <%= content_tag(:small, checkbox_helper_text) unless is_checkbox_disabled %>
```
Smart UI fix here!
